release-tools/keyrotation/key-rotation.py

```diff
@@ -0,0 +1,237 @@
+        logging.info("New Access ID created : " + key_pair.id)
+        print("New Access ID created : " + key_pair.id)
+        update_git_secret(git_repo,git_owner,git_access_id,key_pair.id)
+        update_git_secret(git_repo,git_owner,git_access_secret,key_pair.secret)
+    except:
+        logging.info("Failed to create key for IAM user : " + iam_username)
+        print("Failed to create key for IAM user : " + iam_username)
+
+
+def main():
+    try:
+        logging.basicConfig(filename='keyrotate.log',level=logging.INFO)
+        if len(sys.argv) < 2:
+            print("Please provide the config file as an argument")
+            exit()
+        with open(sys.argv[1]) as source:
+            source_data = json.loads(source.read())
+            git_owner = source_data.get("key-rotate").get("git-owner")
+            logging.info("The GIT owner defined in the config is : " + git_owner)
+            print("GIT owner : " + git_owner)
+            repos = source_data.get("key-rotate").get("odfe-repos")
+            for repo in repos:
+                try:
+                    print(repo)
+                    logging.info("Key rotation is in progress for Repo : " + repo.get("repo"))
+                    logging.info("Repo configuration : ")
+                    logging.info(repo)
+                    print("Repo name: " + repo.get("repo"))
+                    git_repo = repo.get("repo")
+                    logging.info("IAM username: " + repo.get("aws_iam_user"))
+                    print("IAM username: " + repo.get("aws_iam_user"))
+                    num_of_keys = number_of_keys(repo.get("aws_iam_user"))
+                    logging.info("Number of secret access keys for " + repo.get("aws_iam_user") + " : " + str(num_of_keys))
+                    print("Number of secret access keys for " + repo.get("aws_iam_user") + " : " + str(num_of_keys))
+                    if num_of_keys == 0:
+                        logging.info("Number of secret access keys are 0")
+                        print("Number of secret access keys are 0")
+                        create_key(repo.get("aws_iam_user"),git_repo,git_owner,repo.get("git_access_key_id"),repo.get("git_access_secret_key"))
+                    elif num_of_keys == 1:
+                        logging.info("Number of secret access keys are 1")
+                        print("Number of secret access keys are 1")
+                        dict = list_keys(repo.get("aws_iam_user"), 1)
+                        logging.info(dict)
+                        deactivate_key(repo.get("aws_iam_user"),dict.get('access_key_id'))
+                        logging.info("Access key has been deactivated : " + dict.get('access_key_id'))
+                        print("Access key has been deactivated : " + dict.get('access_key_id'))
+                        create_key(repo.get("aws_iam_user"),git_repo,git_owner,repo.get("git_access_key_id"),repo.get("git_access_secret_key"))
```

Comment:
You should create the new key before deactivating the old one, otherwise there's a window where there are no active keys. If key creation fails then there will be no active keys!

Reply:
+1

Reply:
Will make the change and validate.

Reply:
Change made to deactivate the key after successful creation.